bigframes/core/compile/scalar_op_compiler.py

```diff
@@ -0,0 +1,1101 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import functools
+import typing
+
+import ibis
+import ibis.common.exceptions
+import ibis.expr.datatypes as ibis_dtypes
+import ibis.expr.operations.generic
+import ibis.expr.types as ibis_types
+import numpy as np
+import pandas as pd
+
+import bigframes.constants as constants
+import bigframes.dtypes
+import bigframes.dtypes as dtypes
+import bigframes.operations as ops
+
+_ZERO = typing.cast(ibis_types.NumericValue, ibis_types.literal(0))
+_NAN = typing.cast(ibis_types.NumericValue, ibis_types.literal(np.nan))
+_INF = typing.cast(ibis_types.NumericValue, ibis_types.literal(np.inf))
+_NEG_INF = typing.cast(ibis_types.NumericValue, ibis_types.literal(-np.inf))
+
+# Approx Highest number you can pass in to EXP function and get a valid FLOAT64 result
+# FLOAT64 has 11 exponent bits, so max values is about 2**(2**10)
+# ln(2**(2**10)) == (2**10)*ln(2) ~= 709.78, so EXP(x) for x>709.78 will overflow.
+_FLOAT64_EXP_BOUND = typing.cast(ibis_types.NumericValue, ibis_types.literal(709.78))
+
+
+class ScalarOpCompiler:
+    # Mapping of operation name to implemenations
+    _registry: dict[
+        str,
+        typing.Callable[
+            [typing.Sequence[ibis_types.Value], ops.RowOp], ibis_types.Value
+        ],
+    ] = {}
+
+    def compile_row_op(
+        self, op: ops.RowOp, inputs: typing.Sequence[ibis_types.Value]
+    ) -> ibis_types.Value:
+        impl = self._registry[op.name]
+        return impl(inputs, op)
+
+    def register_unary_op(
+        self,
+        op_ref: typing.Union[ops.UnaryOp, type[ops.UnaryOp]],
+        pass_op: bool = False,
```

Comment:
Let's document that the purpose of this is for functions that need to pull out the parameterization from an op instance.